go/internal/feast/logging/logger.go

```diff
@@ -0,0 +1,263 @@
+
+	FeatureValues   []*types.Value
+	FeatureStatuses []serving.FieldStatus
+	EventTimestamps []*timestamppb.Timestamp
+
+	RequestId    string
+	LogTimestamp time.Time
+}
+
+type LogSink interface {
+	Flush() error
+	Write(data arrow.Record) error
+}
+
+type Logger interface {
+	Log(joinKeyToEntityValues map[string][]*types.Value, featureVectors []*serving.GetOnlineFeaturesResponse_FeatureVector, featureNames []string, requestData map[string]*types.RepeatedValue, requestId string) error
+}
+
+type LoggerImpl struct {
+	featureServiceName string
+
+	buffer *MemoryBuffer
+	schema *FeatureServiceSchema
+
+	logCh    chan *Log
+	signalCh chan interface{}
+
+	sink   LogSink
+	config LoggerConfig
+
+	isStopped bool
+	cond      *sync.Cond
+}
+
+type LoggerConfig struct {
+	LoggingOptions
+
+	SampleRate float32
+}
+
+func NewLoggerConfig(sampleRate float32, opts LoggingOptions) LoggerConfig {
+	return LoggerConfig{
+		LoggingOptions: opts,
+		SampleRate:     sampleRate,
+	}
+}
+
+func NewLogger(schema *FeatureServiceSchema, featureServiceName string, sink LogSink, config LoggerConfig) (*LoggerImpl, error) {
+	logger := &LoggerImpl{
+		featureServiceName: featureServiceName,
+
+		logCh:    make(chan *Log, config.ChannelCapacity),
+		signalCh: make(chan interface{}, 2),
+		sink:     sink,
+
+		buffer: &MemoryBuffer{
+			logs:   make([]*Log, 0),
+			schema: schema,
+		},
+		schema: schema,
+		config: config,
+
+		isStopped: false,
+		cond:      sync.NewCond(&sync.Mutex{}),
+	}
+
+	logger.startLoggerLoop()
+	return logger, nil
+}
+
+func (l *LoggerImpl) EmitLog(log *Log) error {
+	select {
+	case l.logCh <- log:
+		return nil
+	case <-time.After(l.config.EmitTimeout):
+		return fmt.Errorf("could not add to log channel with capacity %d. Operation timed out. Current log channel length is %d", cap(l.logCh), len(l.logCh))
+	}
+}
+
+func (l *LoggerImpl) startLoggerLoop() {
+	go func() {
+		for {
+			if err := l.loggerLoop(); err != nil {
+				log.Printf("LoggerImpl[%s] recovered from panic: %+v", l.featureServiceName, err)
+
+				// Sleep for a couple of milliseconds to avoid CPU load from a potential infinite panic-recovery loop
+				time.Sleep(5 * time.Millisecond)
+				continue // try again
+			}
+
+			// graceful stop
+			return
+		}
+	}()
+}
+
+// Select that either ingests new logs that are added to the logging channel, one at a time to add
+// to the in-memory buffer or flushes all of them synchronously to the OfflineStorage on a time interval.
+func (l *LoggerImpl) loggerLoop() (lErr error) {
+	defer func() {
+		// Recover from panic in the logger loop, so that it doesn't bring down the entire feature server
+		if r := recover(); r != nil {
+			rErr, ok := r.(error)
+			if !ok {
+				rErr = fmt.Errorf("%v", r)
+			}
+			lErr = errors.WithStack(rErr)
+		}
+	}()
+	for {
+		shouldStop := false
+
+		select {
+		case <-l.signalCh:
+			l.buffer.writeBatch(l.sink)
+			l.sink.Flush()
+			shouldStop = true
+		case <-time.After(l.config.WriteInterval):
```

Comment:
Maybe add some comments here that specify that differentiates flushing from writing.

Reply:
Added comments to LogSink interface